### PR TITLE
Add annotations and py.typed to conform with PEP561. Add type checking to tox

### PR DESCRIPTION
**Pull request by keiclone**
*Issue number of the reported bug or feature request: #24

**Describe your changes**
made `attrs-strict` PEP561 compatible by adding annotations and `py.typed`. This removes the missing import error from `mypy` when attempting to run static type checking on packages using `attrs-strict`

**Testing performed**
Was able to successfully run static type checking in a test project after installing attrs-strict from my branch.

when testing on a separate library:
```python
# test.py
@attr.s
class FooClass(object):
    list_of_numbers = attr.ib(validator=type_validator(), type=List[int])


fc = FooClass([1, 2, 3, "four"])
```

```bash
# before changes
$ pipenv install attrs-strict
Installing git+https://github.com/bloomberg/attrs-strict
...
$ mypy test.py
error: Cannot find implementation or library stub for module named 'attrs_strict'
note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
error: List item 3 has incompatible type "str"; expected "int"

# after changes
$ pipenv install git+https://github.com/keiclone/attrs-strict#egg=attrs-strict
Installing git+https://github.com/keiclone/attrs-strict#egg=attrs-strict
...
$ mypy test.py
error: List item 3 has incompatible type "str"; expected "int"
```

**Additional context**
could use some extra attention on how this was added to tox to make sure it was done correctly

### CODE REVIEW

**Comment by keiclone** [comment]
@gaborbernat sorry for opening a new PR, I borked my rebase process and figured it'd be easier/less risky to start over.

One thing to note is that it seems mypy seems to have some strange behavior when we do things like compare `typing.Type[Any]` to `typing.Union`, specifically a "Nonoverlapping comparison". Seems like an edge case that we can ignore for now.

I think I addressed all the comments in #31. Could you please take a look?

File: MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1,3 @@
+include attrs_strict/py.typed
```
Comment (gaborbernat) [inline MANIFEST.in:1]: use ``setup.cfg`` and package data to include the py.typed

File: MANIFEST.in [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
+recursive-include attrs_strict *.pyi
```
Comment (gaborbernat) [inline MANIFEST.in:2]: this and the line below is not required with setuptools-scm that pulls in all SCM included. files https://github.com/bloomberg/attrs-strict/blob/master/pyproject.toml#L4

so this file may be removed entirely

File: .gitignore [hunk 3]
```diff
@@ -17,6 +17,7 @@ __pycache__
 /.vscode
 
 # tools
-/.*_cache
+.*_cache
```
Comment (gaborbernat) [inline .gitignore:20]: why is this change needed?

File: attrs_strict/_type_validation.py [hunk 4]
```diff
@@ -19,12 +19,14 @@
...
 except ImportError:
-    from funcsigs import signature
+    # silencing type error so mypy doesn't complain about duplicate import
+    from funcsigs import signature  # type: ignore
 
 try:
     from itertools import zip_longest
 except ImportError:
-    from itertools import izip_longest as zip_longest
+    # silencing type error so mypy doesn't complain about duplicate import
```
Comment (gaborbernat) [inline attrs_strict/_type_validation.py:28]: this feels unneeded verbose; this would suffice in one-line 

```
# type: ignore # duplicate import
```

Consider though the alternative way of expressing this more correctly:
```
if sys.version_info >= (3,8):
  import
else:
  import
```

File: tox.ini [hunk 5]
```diff
@@ -32,6 +34,25 @@ commands = python -m pytest \
...
+basepython = python3.7
+deps = mypy
+setenv =
+    {[testenv]setenv}
+    MYPYPATH={toxinidir}
+commands = mypy types --py2 --strict
+           mypy types --strict
+
+[testenv:merge]
+description = try to merge our types against our source
```
Comment (gaborbernat) [inline tox.ini:49]: I don't see a reason for merge and mypy to not be in the same type tox target environment and merge the files in a tempdir 🤷

File: .gitignore (hunk 3 above)
Comment (keiclone) [reply .gitignore:20]: I think it was because it wasn't ignoring `mypy_cache` before

File: MANIFEST.in (hunk 2 above)
Comment (keiclone) [reply MANIFEST.in:2]: so if I understand you correctly, we just move `py.typed` into `setup.cfg` and remove the rest of this file, and somehow the `*.pyi` files will be included automatically?

File: .gitignore (hunk 3 above)
Comment (keiclone) [reply .gitignore:20]: just tested again, it was only the `mypy_cache` in the `tests` folder, so I'm removing this change

File: src/attrs_strict/_type_validation.py [hunk 6]
```diff
@@ -64,29 +66,34 @@ def _validator(instance, attribute, field):
 
 
 def _validate_elements(attribute, value, expected_type):
+    if expected_type is None:
+        return
+
     base_type = _get_base_type(expected_type)
 
-    if base_type is None or base_type == typing.Any:
+    if base_type == typing.Any:
```
Comment (erikseulean) [inline src/attrs_strict/_type_validation.py:74]: Why is the none check dropped here ?

Comment (gaborbernat) [reply src/attrs_strict/_type_validation.py:74]: Not sure to be fair; I assume given the type checker does not complain it's not needed? @keiclone knows probably better

**Comment by keiclone** [comment]
The None check was effectively moved up to line 69. `_get_base_type` only takes in and returns `typing.Type` types.

**Comment by gaborbernat** [comment]
@erikseulean want to do a new release?

**Comment by erikseulean** [comment]
Yes, feel free to trigger a new one.
